### PR TITLE
Update nullable value to GetValueOrDefault

### PR DESCRIPTION
**Pull request by jwood803**
Update for #2612. Set as WIP to start getting feedback.

@TomFinley Just let me know if I missed anything.

### CODE REVIEW

**Comment by codemzs** [comment]
@jwood803 Can this PR be closed? It has a WIP tag.

**Comment by jwood803** [comment]
@codemzs I added the WIP just for an initial review to make sure the changes were expected. The issue is still open, though, so I can continue to work on this.

**Comment by codemzs** [comment]
@jwood803 Thanks. I can review whenever you think it is ready.

**Comment by jwood803** [comment]
@codemzs Removed WIP tag so this PR should now be good to go.

**Comment by codemzs** [comment]
@jwood803 Thanks, have you verified this change has actually improved the performance? Any benchmark tests?

**Comment by jwood803** [comment]
@codemzs I haven't run any, though @TomFinley mentions this in the issue:

> While I don't expect a measurable perf impact, since I don't think we use nullables in tight loops too often if anywhere, it certainly wouldn't hurt and would help to maintain best practices.

Perhaps I can run the benchmark tests, though, if that works as a before and after?

**Comment by codemzs** [comment]
@jwood803 that would be great and can also please sync your branch with master and update this PR?

File: src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs [hunk 1]
```diff
@@ -371,12 +371,12 @@ private static float CalculateLossSlope(float label, float modelResponse)
...
             columns.Add(data.Schema.Label.Value);
             if (data.Schema.Weight != null)
-                columns.Add(data.Schema.Weight.Value);
+                columns.Add(data.Schema.Weight.GetValueOrDefault());
 
             using (var cursor = data.Data.GetRowCursor(columns))
             {
-                var labelGetter = RowCursorUtils.GetLabelGetter(cursor, data.Schema.Label.Value.Index);
-                var weightGetter = data.Schema.Weight.HasValue ? cursor.GetGetter<float>(data.Schema.Weight.Value) : null;
+                var labelGetter = RowCursorUtils.GetLabelGetter(cursor, data.Schema.Label.GetValueOrDefault().Index);
```
Comment (codemzs) [inline src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs]: >data.Schema.Label [](start = 72, length = 17)

can we check data.Schema.Label has value? Please also take care of line 372

File: src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs [hunk 2]
```diff
@@ -371,12 +371,12 @@ private static float CalculateLossSlope(float label, float modelResponse)
             var columns = new List<DataViewSchema.Column>(featureColumns);
             columns.Add(data.Schema.Label.Value);
```
Comment (codemzs) [inline src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs]: >Value [](start = 42, length = 5)

GetValueOrDefault() and also check before if it has value.

File: src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs [hunk 3]
```diff
@@ -371,12 +371,12 @@ private static float CalculateLossSlope(float label, float modelResponse)
...
             if (data.Schema.Weight != null)
```
Comment (codemzs) [inline src/Microsoft.ML.StandardTrainers/FactorizationMachine/FactorizationMachineTrainer.cs]: >(data.Schema.Weight != null) [](start = 15, length = 28)

lets do data.Schema.Weight.HasValue to keep it consistent.

**Comment by codemzs** [comment]
>             env.CheckUserArg(!options.ModelKind.HasValue || Enum.IsDefined(typeof(WordEmbeddingEstimator.PretrainedModelKind), options.ModelKind), nameof(options.ModelKind));

this seem suspicious. Wouldn't line 226 ensure options.ModelKind will have Value? and check here needs to be options.ModelKind.HasValue == true?

---
Refers to: src/Microsoft.ML.Transforms/Text/WordEmbeddingsExtractor.cs:227 in feeeabf. [](commit_id = feeeabf8d3379a4a8fd1057220afd59294b1134c, deletion_comment = False)

File: src/Microsoft.ML.Transforms/Text/WordEmbeddingsExtractor.cs [hunk 4]
```diff
@@ -241,7 +241,7 @@ internal static IDataTransform Create(IHostEnvironment env, Options options, IDa
             if (customLookup)
                 return new WordEmbeddingTransformer(env, options.CustomLookupTable, cols).MakeDataTransform(input);
             else
-                return new WordEmbeddingTransformer(env, options.ModelKind.Value, cols).MakeDataTransform(input);
+                return new WordEmbeddingTransformer(env, options.ModelKind.GetValueOrDefault(), cols).MakeDataTransform(input);
```
Comment (codemzs) [inline src/Microsoft.ML.Transforms/Text/WordEmbeddingsExtractor.cs]: >options.ModelKind.GetValueOrDefault() [](start = 57, length = 37)

We need to ensure on line 227 that options.ModelKind has Value.

**Comment by codemzs** [comment]
ping.

---
In reply to: [537742508](https://github.com/dotnet/machinelearning/pull/3746#issuecomment-537742508) [](ancestors = 537742508)

**Comment by codemzs** [comment]
ping.

---
In reply to: [537907204](https://github.com/dotnet/machinelearning/pull/3746#issuecomment-537907204) [](ancestors = 537907204)

**Comment by jwood803** [comment]
@codemzs @TomFinley I ran some of the benchmark tests and noticed these results:

Running on `master` branch:
```
                      Method |      Mean |     Error |    StdDev | Extra Metric |
---------------------------- |----------:|----------:|----------:|-------------:|
         MakeIrisPredictions |  6.739 ms | 0.8297 ms | 0.9555 ms |            - |
    MakeSentimentPredictions | 53.858 ms | 0.7493 ms | 0.7009 ms |            - |
 MakeBreastCancerPredictions |  2.280 ms | 0.1301 ms | 0.1498 ms |            - |
```

Running on this branch:
```
                      Method |      Mean |     Error |    StdDev | Extra Metric |
---------------------------- |----------:|----------:|----------:|-------------:|
         MakeIrisPredictions |  7.614 ms | 0.9958 ms | 1.1468 ms |            - |
    MakeSentimentPredictions | 57.186 ms | 1.5118 ms | 1.7410 ms |            - |
 MakeBreastCancerPredictions |  2.428 ms | 0.1833 ms | 0.2111 ms |            - |
```

It actually seems performance is a bit worse with these changes. Not sure if this causes too much of a perf increase to include these changes or not, but I'm curious to what y'alls thoughts are.

**Comment by codemzs** [comment]
@jwood803 thanks for submitting this change. There is this quote "In theory there is no difference between theory and practice. But, in practice, there is", thank you for benchmarking this change. Can you please make sure your branch was synced to the same commit that master was synced to when you benchmarked? Some of these results seem significant but we should try to increase the dataset size or may be the run the pipeline in a loop to see if this difference actually increases with more processing. Would be nice if you can also try that?

**Comment by codemzs** [comment]
@jwood803 I haven't heard back from you in a while so I will close this PR for now but thank you for doing all the hard work to make this change and later profiling to make sure performance has not regressed. Its unfortunate this change does not really improve the performance as we expected but instead does the opposite.

**Comment by jwood803** [comment]
Thanks @codemzs! I appreciate the time reviewing the PR even if it doesn't seem like this improves the performance. Should we close the issues as well stating this?

**Comment by codemzs** [comment]
@jwood803 Yes, please close the issue and no worries! :)

**Pull request closed without merging**